File: apps/admin-web/src/infrastructure/api/diningApi.ts
import { apiRequest } from './httpClient';

const DINING_SERVICE_URL = import.meta.env.VITE_DINING_SERVICE_URL;

export type TableStatus = 'AVAILABLE' | 'OCCUPIED' | 'RESERVED' | 'WAITING_CLEANING';
export type CommandStatus = 'OPEN' | 'CLOSED';

export interface Table {
  id: string;
  tenant_id: string;
  number: number;
  capacity: number;
  status: TableStatus;
  qr_code_url: string;
}

export interface Command {
  id: string;
  tenant_id: string;
  table_id: string;
  customer_name: string;
  customer_cpf: string | null;
  waiter_id: string;
  status: CommandStatus;
  service_fee_charged: boolean;
  opened_at: string;
  closed_at: string | null;
}

export interface CreateTableRequest {
  number: number;
  capacity: number;
  qr_code_url?: string;
}

export async function listTables(): Promise<Table[]> {
  return apiRequest<Table[]>(DINING_SERVICE_URL, '/api/v1/dining/tables');
}

/** Libera uma mesa "Aguardando Limpeza" de volta pra disponível (US-03.5) —
 * sem isso, toda comanda encerrada prende a mesa pra sempre. */
export async function markTableCleaned(tableId: string): Promise<Table> {
  return apiRequest<Table>(DINING_SERVICE_URL, `/api/v1/dining/tables/${tableId}/mark-cleaned`, {
    method: 'POST',
  });
}

export async function createTable(payload: CreateTableRequest): Promise<Table> {
  return apiRequest<Table>(DINING_SERVICE_URL, '/api/v1/dining/tables', {
    method: 'POST',
    body: { qr_code_url: '', ...payload },
  });
}

export async function listCommands(status?: CommandStatus): Promise<Command[]> {
  const suffix = status ? `?status=${status}` : '';
  return apiRequest<Command[]>(DINING_SERVICE_URL, `/api/v1/dining/commands${suffix}`);
}

export async function setCommandServiceFee(commandId: string, charged: boolean): Promise<Command> {
  return apiRequest<Command>(DINING_SERVICE_URL, `/api/v1/dining/commands/${commandId}/service-fee`, {
    method: 'PATCH',
    body: { charged },
  });
}

export async function closeCommand(commandId: string): Promise<Command> {
  return apiRequest<Command>(DINING_SERVICE_URL, `/api/v1/dining/commands/${commandId}/close`, {
    method: 'POST',
  });
}

export interface TableQrSecret {
  secret: string;
  expires_at: string;
}

export async function rotateTableQrSecret(tableId: string): Promise<TableQrSecret> {
  return apiRequest<TableQrSecret>(DINING_SERVICE_URL, `/api/v1/dining/tables/${tableId}/qr-secret/rotate`, {
    method: 'POST',
  });
}

File: apps/waiter-mobile/lib/data/api/dining_api.dart
import '../dto/command_dto.dart';
import '../dto/table_dto.dart';
import 'api_client.dart';

class DiningApi {
  const DiningApi(this._client);

  final ApiClient _client;

  Future<List<TableDto>> listTables() async {
    final json = await _client.getList('/api/v1/dining/tables');
    return json.map((e) => TableDto.fromJson(e as Map<String, dynamic>)).toList();
  }

  Future<CommandDto> openCommand({
    required int tableNumber,
    required String customerName,
    String? customerCpf,
  }) async {
    final json = await _client.post(
      '/api/v1/dining/commands/open',
      body: {
        'table_number': tableNumber,
        'customer_name': customerName,
        'customer_cpf': ?customerCpf,
      },
    );
    return CommandDto.fromJson(json);
  }

  /// Comandas abertas do tenant — usado para achar o `command_id` de uma
  /// mesa já ocupada (o mapa de mesas só devolve o status, não a comanda).
  Future<List<CommandDto>> listOpenCommands() async {
    final json = await _client.getList('/api/v1/dining/commands?status=OPEN');
    return json.map((e) => CommandDto.fromJson(e as Map<String, dynamic>)).toList();
  }

  /// Libera uma mesa "Aguardando Limpeza" de volta pra disponível (US-03.5)
  /// — sem isso, toda comanda encerrada prende a mesa pra sempre.
  Future<TableDto> markTableCleaned(String tableId) async {
    final json = await _client.post('/api/v1/dining/tables/$tableId/mark-cleaned');
    return TableDto.fromJson(json);
  }
}
